### ISSUE TITLE
False positive error E0277 when trying to implement PartialOrd trait

### ISSUE THREAD

**Issue by only-rust** [intellij-rust/intellij-rust#8583]
## Environment

* **IntelliJ Rust plugin version:** 0.4.165.4438-213
* **Rust toolchain version:** stable-x86_64-pc-windows-msvc @ 1.58.1
* **IDE name and version:** IntelliJ IDEA 2021.3.2 (Community Edition)
* **Operating system:** Windows 7

## Problem description
This is valid rust code that compiles without a problem, but Intellij thinks it is invalid.

![iss](https://user-images.githubusercontent.com/100077089/154900131-d490111f-7a52-4032-abb6-593f9421a94e.png)

## Steps to reproduce
```rust

use std::cmp::Ordering;

#[derive(Debug, PartialEq)]
struct House {
    area: u64,
}

#[derive(Debug, PartialEq)]
struct Village {
    area: u64,
}

impl PartialEq<Village> for House {
    fn eq(&self, other: &Village) -> bool {
        self.area == other.area
    }
}

impl PartialEq<House> for Village {
    fn eq(&self, other: &House) -> bool {
        self.area == other.area
    }
}

impl PartialOrd<Village> for House {
    fn partial_cmp(&self, other: &Village) -> Option<Ordering> {
        Some(self.area.cmp(&other.area))
    }
}

impl PartialOrd<House> for Village {
    fn partial_cmp(&self, other: &House) -> Option<Ordering> {
        Some(self.area.cmp(&other.area))
    }
}

```
